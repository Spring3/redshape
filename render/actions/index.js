import issueActions from './issue.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  issue: issueActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
