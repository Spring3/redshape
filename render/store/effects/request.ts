import { ipcRenderer, remote } from 'electron';
import { Response } from '../../../types';

const crypto = remote.require('crypto');

type Config = {
  endpoint: string;
  token?: string;
}

type QueryConfig = {
  payload: Record<string, any>;
  config?: Config;
}

const query = (config: QueryConfig): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('request', JSON.stringify({ ...config, id }));

    ipcRenderer.once(`response:${id}`, (event, response) => {
      console.log('response:', response);
      resolve(response);
    });
  });
};

export {
  query,
};

export type {
  QueryConfig
};
