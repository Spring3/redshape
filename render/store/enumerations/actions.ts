import type { IAction, Context } from 'overmind';
import { Activity, Response } from '../../../types';

const fetchActivities: IAction<void, Promise<Response<Activity[]>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'enumerations/time_entry_activities.json',
    }
  });

  if (response.success) {
    state.enumerations.activities = response.data.activities as Activity[] ?? [];
    return {
      success: true,
      data: state.enumerations.activities,
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  fetchActivities
};
