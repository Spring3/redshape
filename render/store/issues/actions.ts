import type { Context, IAction } from 'overmind';
import { Issue, PaginatedActionResponse, Response } from '../../../types';
import { indexById } from '../helpers';

type GetManyIssuesArgs = {
  filters: {
    assigned_to_id?: number | string;
    status_id?: '*' | 'open' | 'closed';
    subject?: string;
    sort?: string;
  },
  offset?: number;
  limit?: number;
}

const getMany: IAction<GetManyIssuesArgs, Promise<PaginatedActionResponse<Issue>>> = async ({ effects, state }: Context, { filters, offset = 0, limit = 20 }) => {
  const query = {
    limit: limit ? Math.abs(limit) : undefined,
    offset: offset ? Math.abs(offset) : 0,
    include: 'attachments,relations',
    ...filters
  };

  const response = await effects.mainProcess.request({
    payload: {
      route: 'issues.json',
      method: 'GET',
      query
    }
  });

  if (response.success) {
    state.issues.byId = {
      ...state.issues.byId,
      ...indexById(response.payload.issues)
    };

    console.log('payload', response.payload);

    return {
      success: true,
      data: {
        items: response.payload.issues,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issues.length),
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit,
      offset
    },
    hasMore: false,
    error: response.error
  };
};

type GetOneIssueArgs = {
  id: string;
}

const getOne: IAction<GetOneIssueArgs, Promise<Response<Issue>>> = async ({ effects, state }: Context, { id }) => {
  const response = await effects.mainProcess.request({
    payload: {
      route: `issues/${id}.json`,
      method: 'GET',
      query: {
        include: 'attachments,children,relations,journals',
      }
    }
  });

  if (response.success) {
    const issue = response.payload as Issue;
    state.issues.byId[id] = issue;
    return {
      success: true,
      data: issue
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  getOne,
  getMany
};
