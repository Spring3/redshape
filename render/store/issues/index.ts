import { state, IssuesState } from './state';
import * as actions from './actions';

export {
  state,
  actions
};

export type {
  IssuesState
};
