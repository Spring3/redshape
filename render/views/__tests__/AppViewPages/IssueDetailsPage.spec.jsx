import React from 'react';
import moment from 'moment';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { HashRouter } from 'react-router-dom';
import MockAdapter from 'axios-mock-adapter';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import theme from '../../../theme';
import IssueDetailsPage from '../../AppViewPages/IssueDetailsPage';
import { ISSUE_RESET } from '../../../actions/issue.actions';
import { ISSUES_GET, ISSUES_RESET_SELECTION } from '../../../actions/issues.actions';
import { getInstance, reset, initialize } from '../../../../common/request';

const mockStore = configureStore([thunk]);
const redmineEndpoint = 'redmine.test.test';
const token = 'multipass';

let axiosMock;

const state = {
  projects: {
    data: {
      1: {
        activities: [
          { id: 1, name: 'Development' },
          { id: 2, name: 'Testing' }
        ]
      }
    }
  },
  user: {
    id: 1,
    name: 'John Wayne'
  },
  issue: {
    isFetching: false,
    error: undefined
  },
  issues: {
    selected: {
      data: {
        id: 1,
        subject: 'Test',
        journals: [],
        description: 'Fixture test issue',
        priority: {
          id: 1,
          name: 'Low'
        },
        project: {
          id: 1,
          name: 'Project 1'
        },
        author: {
          id: 1,
          name: 'Author 1'
        },
        done_ratio: 0.5,
        start_date: '2011-01-01',
        due_date: '2011-01-01',
        total_estimated_hours: 15,
        spent_hours: 2,
        status: {
          id: 1,
          name: 'Open'
        },
        tracker: {
          id: 1,
          name: 'Issue'
        },
        assigned_to: {
          id: 1,
          name: 'John Wayne'
        }
      },
      spentTime: {
        data: [],
        isFetching: false,
        error: undefined
      }
    }
  },
  tracking: {
    issue: {
      id: 2
    },
    isEnabled: false,
    duration: 0
  },
  settings: {
    idleBehavior: 0,
    discardIdleTime: true,
    advancedTimerControls: false,
    progressWithStep1: false,
  }
};

describe('IssueDetails page', () => {
  beforeAll(() => {
    initialize(redmineEndpoint, token);
    axiosMock = new MockAdapter(getInstance());
  });

  afterEach(() => {
    axiosMock.reset();
  });

  afterAll(() => {
    axiosMock.restore();
    reset();
  });

  it('should match the snapshot', () => {
    axiosMock.onAny().reply(() => Promise.resolve([200, {}]));
    const store = mockStore(state);
    const tree = renderer.create(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssueDetailsPage
              match={{
                params: {
                  id: 1
                }
              }}
              history={{
                goBack: jest.fn()
              }}
            />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fetch the info about a selected issue', () => {
    axiosMock.onAny().reply(() => Promise.resolve([200, {}]));
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssueDetailsPage
              match={{
                params: {
                  id: 1
                }
              }}
              history={{
                goBack: jest.fn()
              }}
            />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );

    expect(store.getActions()[0].type).toBe(ISSUES_GET);
  });

  it('should reset the selected issue on unmount', () => {
    axiosMock.onAny().reply(() => Promise.resolve([200, {}]));
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssueDetailsPage
              match={{
                params: {
                  id: 1
                }
              }}
              history={{
                goBack: jest.fn()
              }}
            />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );

    wrapper.unmount();

    const actions = store.getActions();
    actions.pop()
    expect(actions.pop().type).toBe(ISSUES_RESET_SELECTION);
  });

  it('should show the selected time entry in a modal', () => {
    axiosMock.onAny().reply(() => Promise.resolve([200, {}]));
    const store = mockStore({
      ...state,
      timeEntry: {}
    });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssueDetailsPage
              match={{
                params: {
                  id: 1
                }
              }}
              history={{
                goBack: jest.fn()
              }}
            />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );

    const page = wrapper.find('IssueDetailsPage').instance();
    expect(page).toBeTruthy();
    const timeEntry = {
      user: {
        id: 1,
        name: 'John Wayne'
      },
      issue: {
        id: 1
      },
      activity: {
        id: 1,
        name: 'Testing'
      },
      project: {
        id: 1,
        name: 'Test'
      },
      hours: 1.5,
      spent_on: '2011-01-01'
    };

    page.showTimeEntryModal(timeEntry);

    expect(page.state.showTimeEntryModal).toBe(true);
    expect(page.state.selectedTimeEntry).toEqual(
      Object.assign({}, timeEntry, {
        issue: {
          id: 1,
          name: state.issues.selected.data.subject
        }
      })
    );
    expect(page.state.activities).toEqual([
      { value: 1, label: 'Development' },
      { value: 2, label: 'Testing' }
    ]);
  });

  it('should show a template for the time entry in a modal', () => {
    axiosMock.onAny().reply(() => Promise.resolve([200, {}]));
    const store = mockStore({
      ...state,
      timeEntry: {}
    });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssueDetailsPage
              match={{
                params: {
                  id: 1
                }
              }}
              history={{
                goBack: jest.fn()
              }}
            />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );

    const page = wrapper.find('IssueDetailsPage').instance();
    expect(page).toBeTruthy();
    const timeEntry = {
      user: {
        id: state.user.id,
        name: state.user.name
      },
      issue: {
        id: state.issues.selected.data.id
      },
      activity: {},
      project: {
        id: state.issues.selected.data.project.id,
        name: state.issues.selected.data.project.name
      },
      duration: '',
      hours: undefined,
      spent_on: moment().format('YYYY-MM-DD')
    };

    page.showTimeEntryModal();

    expect(page.state.showTimeEntryModal).toBe(true);
    expect(page.state.selectedTimeEntry).toEqual(
      Object.assign({}, timeEntry, {
        issue: {
          id: 1,
          name: state.issues.selected.data.subject
        }
      })
    );
    expect(page.state.activities).toEqual([
      { value: 1, label: 'Development' },
      { value: 2, label: 'Testing' }
    ]);
  });

  it('should close the time entry modal', () => {
    axiosMock.onAny().reply(() => Promise.resolve([200, {}]));
    const store = mockStore({
      ...state,
      timeEntry: {}
    });
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssueDetailsPage
              match={{
                params: {
                  id: 1
                }
              }}
              history={{
                goBack: jest.fn()
              }}
            />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );

    const page = wrapper.find('IssueDetailsPage').instance();
    expect(page).toBeTruthy();

    page.showTimeEntryModal();
    expect(page.state.showTimeEntryModal).toBe(true);

    page.closeTimeEntryModal();
    expect(page.state.showTimeEntryModal).toBe(false);

    expect(page.state.selectedTimeEntry).toBe(undefined);
    expect(page.state.activities).toEqual([]);
  });
});
