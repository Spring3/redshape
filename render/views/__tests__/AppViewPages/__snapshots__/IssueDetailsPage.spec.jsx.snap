// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueDetails page should match the snapshot 1`] = `
.c20 {
  position: absolute;
  display: none;
  border-radius: 3px;
  width: auto;
  padding: 5px;
  background: #FFFFFF;
  color: #EF6069;
  text-align: center;
  font-size: 12px;
  white-space: nowrap;
  box-shadow: 0px 2px 7px #A0A0A0;
  z-index: 99;
  margin-block-start: 1em;
  margin-block-end: 1em;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  bottom: 75%;
  left: 50%;
}

.c20::after {
  content: ' ';
  position: absolute;
  width: auto;
  border: 2px solid black;
  border-width: 5px;
  border-color: #FFFFFF transparent transparent transparent;
}

.c20::after {
  top: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c16 {
  display: inline-block;
  position: relative;
}

.c16:hover .c19 {
  display: block;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c17 {
  background: #F9F9F9;
  position: relative;
  border-radius: 5px;
  height: 7px;
  width: 150px;
}

.c18 {
  border-radius: 5px;
  max-width: 100%;
  width: 0;
  float: none;
  -webkit-transition: width ease;
  transition: width ease;
  width: 0.5% !important;
  height: 7px;
  background: #FF634D;
}

.c21 {
  border-radius: 5px;
  max-width: 100%;
  width: 0;
  float: none;
  -webkit-transition: width ease;
  transition: width ease;
  width: 0% !important;
  height: 7px;
  background: #6CCA51;
}

.c10 {
  border-radius: 3px;
  font-weight: bold;
  font-size: 14px;
  outline: none;
  text-align: center;
  background: #FFFFFF;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
  width: auto;
  border: 2px solid #FF7079;
  color: #FF7079;
  cursor: pointer;
}

.c10:hover,
.c10:focus {
  background: #FF7079;
  color: #FFFFFF !important;
}

.c10:hover svg,
.c10:focus svg {
  fill: #FFFFFF;
}

.c10:active {
  background: #EF6069;
}

.c22 {
  box-sizing: border-box;
  overflow: hidden;
}

.c12 {
  position: relative;
  display: inline-grid;
  grid-template-rows: 1fr;
}

.c12 span {
  float: left;
  grid-row: 1;
  grid-column: 1;
}

.c12 span:last-child {
  opacity: 0;
  display: none;
}

.c24 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-radius: 3px;
}

.c24 h2 {
  display: inline-block;
  margin-left: 15px;
  margin-right: 15px;
}

.c24 > div {
  margin-right: 15px;
  padding: 0px 5px;
  border-radius: 3px;
}

.c24 > div button {
  background: #FFFFFF;
  margin: 0px 5px;
}

.c25 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FF7F87;
}

.c23 {
  background: white;
  max-height: 550px;
  min-width: 350px;
}

.c26 {
  position: relative;
  list-style-type: none;
  border-radius: 3px;
  margin: 0;
  overflow-y: scroll;
  max-height: 500px;
  background: #F9F9F9;
  box-shadow: inset 0px 0px 10px 0px #F9F9F9;
  padding: 0px 10px 10px 10px;
}

.c26 li {
  cursor: pointer;
  display: block;
  padding: 10px;
  border-radius: 3px;
  border: 1px solid #E9E9E9;
  background: #FFFFFF;
  margin: 10px auto 0px auto;
}

.c26 li div.status {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c26 li div.status a {
  margin-left: 10px;
  visibility: hidden;
}

.c26 li div.status div {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c26 li div.status div span.time,
.c26 li div.status div span.date {
  margin-right: 5px;
  font-size: 12px;
  color: #A0A0A0;
}

.c26 li div.status div span.username {
  font-weight: bold;
  margin-right: 5px;
  color: #3F3844;
}

.c26 li div.extra {
  display: grid;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  grid-template-columns: 1fr auto;
  grid-auto-flow: row;
  grid-column-gap: 1rem;
  grid-row-gap: 3px;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
}

.c26 li div.extra:not(:empty) {
  margin-top: 0.5rem;
}

.c26 li div.extra span {
  font-size: 12px;
  color: #A0A0A0;
}

.c26 li div.extra span:nth-child(even) {
  text-align: right;
}

.c26 li p {
  margin-bottom: 0px;
  min-width: 100%;
  width: 0;
  line-height: 2;
}

.c26 li:hover div:first-child a {
  visibility: visible;
}

.c27 {
  background: white;
  padding: 20px;
  margin-bottom: 20px;
}

.c30 {
  list-style-type: none;
  padding: 20px 0px;
  margin: 0px;
  border-radius: 3px;
  background: #F9F9F9;
}

.c30 li:first-child {
  margin-top: 20px;
}

.c30 li {
  margin: 40px 20px 20px 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c30 li div.commentsHeader {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 20%;
}

.c30 li div.commentsHeader h3 {
  margin-top: 20px;
  color: #FF7079;
}

.c30 li div.commentsHeader span:last-child {
  color: #A0A0A0;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  text-align: left;
}

.c30 li div.commentsHeader span:last-child:hover {
  color: #3F3844;
}

.c30 li iframe {
  margin-left: 20px;
  padding: 5px 20px 0px 20px;
  background: #FFFFFF;
  border: 1px solid #E9E9E9;
  border-radius: 3px;
  min-height: 100px;
  width: 74%;
}

.c28 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c28 div:last-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c28 div:last-child button {
  margin-right: 0.5rem;
}

.c28 div:last-child button:last-child {
  margin-right: 0;
}

.c29 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 20px;
}

.c4 {
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
}

.c0 {
  background: white;
  padding: 20px;
  margin-bottom: 60px;
}

.c9 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  font-size: 12px;
  margin-top: 0px;
  color: #A0A0A0;
  margin-block-start: 1em;
  margin-block-end: 1em;
}

.c6 a {
  font-size: inherit !important;
  margin: 0 5px;
}

.c8 {
  display: block;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0.95rem 1rem 0 1rem;
}

.c8 a:first-child {
  margin-right: 2rem;
}

.c2 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c14 {
  margin-right: 0.5rem;
}

.c13 {
  color: gray;
}

.c7 {
  color: #FF7079;
  padding: 0 0.3rem;
}

.c11 {
  display: grid;
  grid-template-columns: auto minmax(auto,340px) auto 1fr;
  grid-row-gap: 5px;
  grid-column-gap: 30px;
  margin-right: 30px;
}

.c11 > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 35px;
}

.c11 > div:nth-child(odd) {
  font-weight: bold;
}

@media (min-width:1600px) {
  .c23 {
    min-width: 500px;
  }
}

@media (max-width:1400px) {
  .c11 {
    grid-template-columns: auto minmax(auto,250px) auto 1fr;
    grid-column-gap: 20px;
    margin-right: 20px;
  }
}

@media (max-width:1150px) {
  .c11 {
    grid-template-columns: auto 1fr;
  }
}

<section
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <h2
            className="c5"
          >
            <span>
              #1
              Â 
            </span>
            <span>
              Test
            </span>
          </h2>
          <div
            className="c6"
          >
            Created
             
            by
            <span
              className="c7"
            >
              Author 1
            </span>
          </div>
        </div>
        <div
          className="buttons c8"
        >
          <button
            className="c9 c10"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={22}
              viewBox="0 0 24 24"
              width={22}
            >
              <path
                d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
              />
            </svg>
            <span>
              Â Edit
            </span>
          </button>
        </div>
      </div>
      <div
        className="c11"
      >
        <div>
          Tracker: 
        </div>
        <div>
          Issue
        </div>
        <div>
          Target version: 
        </div>
        <div />
        <div>
          Status:
        </div>
        <div>
          Open
        </div>
        <div>
          Start date: 
        </div>
        <div>
          <span
            className="c12"
          >
            <span>
              Jan 01 2011
            </span>
            <span>
              Jan 01 2011
            </span>
          </span>
        </div>
        <div>
          Priority: 
        </div>
        <div>
          Low
        </div>
        <div>
          Due date: 
        </div>
        <div>
          <span
            className="c12"
          >
            <span>
              Jan 01 2011
            </span>
            <span>
              Jan 01 2011
            </span>
          </span>
        </div>
        <div>
          Assignee: 
        </div>
        <div>
          John Wayne
        </div>
        <div>
          Estimation: 
        </div>
        <div>
          <span
            className="c13"
          >
            (Total: 15.00 h)
          </span>
        </div>
        <div>
          Project: 
        </div>
        <div>
          Project 1
        </div>
        <div>
          Time spent: 
        </div>
        <div>
          <span
            className="c14"
          >
            2.00 h
          </span>
        </div>
        <div>
          Progress: 
        </div>
        <div>
          <div
            className="c15"
          >
            <div
              className="c16"
            >
              <div
                className="c17"
                height={7}
                width={150}
              >
                <div
                  className="c18"
                  height={7}
                />
              </div>
              <div
                className="tooltip c19 c20"
              >
                1%
              </div>
            </div>
          </div>
        </div>
        <div>
          Time cap: 
        </div>
        <div>
          <div
            className="c15"
          >
            <div
              className="c16"
            >
              <div
                className="c17"
                height={7}
                width={150}
              >
                <div
                  className="c21"
                  height={7}
                />
              </div>
              <div
                className="tooltip c19 c20"
              >
                0%
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <h3>
          Description
        </h3>
        <iframe
          className="c22"
          frameBorder="0"
          onLoad={[Function]}
          srcDoc="<p>Fixture test issue</p>"
          width="100%"
        />
      </div>
    </div>
    <div
      className="c23"
    >
      <div
        className="c24"
      >
        <h2
          className="padded"
        >
          Time spent
        </h2>
        <div>
          <button
            className="c25 c10"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={22}
              viewBox="0 0 24 24"
              width={22}
            >
              <path
                d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
              />
            </svg>
            <span>
              Â Add
            </span>
          </button>
          <button
            className="c25 c10"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={22}
              viewBox="0 0 24 24"
              width={22}
            >
              <path
                d="M12,20A7,7 0 0,1 5,13A7,7 0 0,1 12,6A7,7 0 0,1 19,13A7,7 0 0,1 12,20M19.03,7.39L20.45,5.97C20,5.46 19.55,5 19.04,4.56L17.62,6C16.07,4.74 14.12,4 12,4A9,9 0 0,0 3,13A9,9 0 0,0 12,22C17,22 21,17.97 21,13C21,10.88 20.26,8.93 19.03,7.39M11,14H13V8H11M15,1H9V3H15V1Z"
              />
            </svg>
            <span>
              Â Track
            </span>
          </button>
        </div>
      </div>
      <ul
        className="c26"
      />
    </div>
  </div>
  <section
    className="c27"
  >
    <div>
      <div
        className="c28"
      >
        <h2>
          Comments
        </h2>
        <div>
          <button
            className="c29 c10"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={22}
              viewBox="0 0 24 24"
              width={22}
            >
              <path
                d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
              />
            </svg>
            Â Editor
          </button>
          <button
            className="c29 c10"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              className="mdi-icon "
              fill="currentColor"
              height={22}
              viewBox="0 0 24 24"
              width={22}
            >
              <path
                d="M10 11V13H18V11H10M10 17V19H14V17H10M10 5V7H22V5H10M6 7H8.5L5 3.5L1.5 7H4V20H6V7Z"
              />
            </svg>
          </button>
        </div>
      </div>
      <ul
        className="c30"
      />
    </div>
  </section>
</section>
`;
